File: packages/android/src/main/kotlin/gov/civui/components/CivFormGroup.kt
// CivUI — CivFormGroup for Jetpack Compose
// Simple wrapper that provides label, hint, and error around arbitrary content.
// Renders: label -> hint -> error -> content (Section 508 compliant)

package gov.civui.components

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.i18n.CivLocale
import gov.civui.tokens.CivTokens

/**
 * Accessible form group wrapper for government applications.
 *
 * Provides label, hint, and error rendering around arbitrary
 * content. Use this to wrap custom or third-party form controls
 * that need consistent CivUI labeling and error patterns.
 *
 * Usage:
 * ```kotlin
 * CivFormGroup(
 *     label = "Custom control",
 *     hint = "Select an option",
 *     error = errorMessage,
 *     required = true,
 * ) {
 *     // Your custom composable here
 * }
 * ```
 */
@Composable
fun CivFormGroup(
    label: String,
    modifier: Modifier = Modifier,
    hint: String? = null,
    error: String? = null,
    required: Boolean = false,
    content: @Composable () -> Unit,
) {
    val isDark = isSystemInDarkTheme()

    val labelColor = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_

    Column(
        modifier = modifier.padding(bottom = CivTokens.Spacing._4),
    ) {
        // 1. Label
        CivLabel(
            label = label,
            required = required,
            labelColor = labelColor,
            errorColor = errorColor,
        )

        // 2. Hint
        CivHint(text = hint, color = hintColor)

        // 3. Error
        CivError(text = error, color = errorColor)

        // 4. Content
        content()
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivFormGroup")
@Composable
private fun CivFormGroupPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var value by remember { mutableStateOf("") }

        CivFormGroup(
            label = "Custom input",
            hint = "This wraps a custom composable",
            required = true,
        ) {
            CivTextInput(
                label = "",
                value = value,
                onValueChange = { value = it },
                placeholder = "Wrapped input",
            )
        }

        CivFormGroup(
            label = "With error",
            error = "Something went wrong",
        ) {
            CivTextInput(
                label = "",
                value = "",
                onValueChange = {},
            )
        }
    }
}

File: packages/ios/Sources/CivUI/CivFormGroup.swift
// CivUI — CivFormGroup for SwiftUI
// Accessible form group wrapper following government design system patterns.
// Renders: label → hint → error → content (Section 508 compliant)

import SwiftUI

/// Accessible form group wrapper for government applications.
///
/// Provides label, hint, and error rendering for custom or third-party
/// form controls. Uses `@ViewBuilder` for arbitrary content.
///
/// Usage:
/// ```swift
/// CivFormGroup(
///     label: "Custom rating",
///     hint: "Drag the slider to rate your experience",
///     error: ratingError
/// ) {
///     Slider(value: $rating, in: 1...5, step: 1)
/// }
/// ```
public struct CivFormGroup<Content: View>: View {
    // MARK: - Properties

    /// Visible label text.
    public let label: String

    /// Help text shown below the label.
    public var hint: String?

    /// Error message.
    public var error: String?

    /// Whether the field is required.
    public var isRequired: Bool

    /// Content rendered inside the form group.
    public let content: Content

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        label: String,
        hint: String? = nil,
        error: String? = nil,
        isRequired: Bool = false,
        @ViewBuilder content: () -> Content
    ) {
        self.label = label
        self.hint = hint
        self.error = error
        self.isRequired = isRequired
        self.content = content()
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._1) {
            // 1. Label
            labelView

            // 2. Hint
            if let hint, !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .accessibilityIdentifier("civ-hint")
            }

            // 3. Error
            if let error, !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityIdentifier("civ-error")
                    .accessibilityAddTraits(.updatesFrequently)
            }

            // 4. Content
            content
        }
        .padding(.bottom, CivTokens.Spacing._4)
        .accessibilityElement(children: .contain)
    }

    // MARK: - Subviews

    private var labelView: some View {
        HStack(spacing: CivTokens.Spacing._0_5) {
            Text(label)
                .font(.system(size: CivTokens.Typography.FontSize.base,
                              weight: CivTokens.Typography.FontWeight.bold))
                .foregroundColor(adaptiveColor(
                    light: CivTokens.Colors.Base.darkest,
                    dark: CivTokens.DarkColors.Base.darkest
                ))

            if isRequired {
                Text("*")
                    .font(.system(size: CivTokens.Typography.FontSize.base,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityLabel("required")
            }
        }
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivFormGroup_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var rating: Double = 3

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivFormGroup(
                        label: "Experience rating",
                        hint: "Drag the slider to rate your experience",
                        isRequired: true
                    ) {
                        Slider(value: $rating, in: 1...5, step: 1)
                        Text("Rating: \(Int(rating))")
                            .font(.system(size: CivTokens.Typography.FontSize.sm))
                    }

                    CivFormGroup(
                        label: "With error",
                        error: "Please provide a rating"
                    ) {
                        Slider(value: .constant(0), in: 1...5, step: 1)
                    }
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif
